src/expand.rs

```diff
@@ -95,9 +102,236 @@ fn impl_struct(input: &DeriveInput, data: &DataStruct) -> Result<TokenStream> {
+) {
+    let display_path = join_paths(&["core", "fmt", "Display"], false);
+
+    let display_bound = TypeParamBound::Trait(TraitBound {
+        paren_token: None,
+        modifier: TraitBoundModifier::None,
+        lifetimes: None,
+        path: display_path,
+    });
+    if !predicate_that_needs_a_display_impl.bounds.is_empty() {
+        predicate_that_needs_a_display_impl.bounds.push_punct(Add {
+            spans: [Span::call_site()],
+        });
+    }
+
+    predicate_that_needs_a_display_impl
+        .bounds
+        .push_value(display_bound);
+}
+
+/// Map each declared generic type parameter to the set of all trait boundaries declared on it.
+///
+/// These boundaries may come from the declaration site:
+///     pub enum E<T: MyTrait> { ... }
+/// or a `where` clause after the parameter declarations:
+///     pub enum E<T> where T: MyTrait { ... }
+/// This method will return the boundaries from both of those cases.
+fn extract_trait_constraints_from_source(
+    where_clause: &WhereClause,
+    type_params: &[&TypeParam],
+) -> Vec<(Ident, Vec<TraitBound>)> {
+    let mut param_constraint_mapping: Vec<(Ident, Vec<TraitBound>)> = type_params
+        .iter()
+        .map(|type_param| {
+            let trait_bounds: Vec<TraitBound> = type_param
+                .bounds
+                .iter()
+                .flat_map(|bound| match bound {
+                    TypeParamBound::Trait(trait_bound) => Some(trait_bound),
+                    _ => None,
+                })
+                .cloned()
+                .collect();
+            (type_param.ident.clone(), trait_bounds)
+        })
+        .collect();
+
+    for predicate in where_clause.predicates.iter() {
+        match predicate {
+            WherePredicate::Type(ref pred_ty) => {
+                let known_bounds: &mut Vec<TraitBound> = match &pred_ty.bounded_ty {
+                    Type::Path(TypePath { path, qself: None }) => {
+                        assert!(
+                            path.leading_colon.is_none(),
+                            "LHS of `where` boundary can't resolve from global scope!"
+                        );
+                        assert_eq!(
+                            1,
+                            path.segments.len(),
+                            "LHS of `where` boundary can't have segments!"
+                        );
+                        let cur_segment = path.segments.first().unwrap();
+                        assert_eq!(PathArguments::None, cur_segment.arguments);
+                        let (_, ref mut known_bounds) = param_constraint_mapping
+                            .iter_mut()
+                            .find(|(id, _)| *id == cur_segment.ident)
+                            .expect("expected all types in `where` boundaries to be declared!");
+                        known_bounds
+                    }
+                    x => unreachable!("expected simple type in where clause, but got {:?}", x),
```

Comment:
I don't think I want to dump `syn` debug data on an error to users.
```suggestion
                    x => unreachable!("encountered unexpected predicate kind"),
```

Reply:
This was a really helpful comment, but I think I have obviated it after spending some some to understand `where` clauses and realizing we don't need to care about anything that's not a direct reference to a type parameter, aka a single `Ident`. Let me know if the result makes sense!